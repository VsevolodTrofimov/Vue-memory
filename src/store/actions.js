import config from '@/src/utility/gameConfig'
import { preloadImages } from '@/src/utility/preload'

export const loadDeck = ({commit, dispatch}) => {
  const loading = preloadImages(config.cards.options)
  commit('deckLoaderRegister', loading)

  return new Promise((resovle, reject) => {
    loading
      .then(() => {
        commit('deckLoaded')
        resovle()
      })
      .catch(reject)
  })
}

export const startGame = ({commit, dispatch, getters}) => {
  if(getters.deck.loaded) {
    commit('fillBoard')
    commit('setScore', 0)
    commit('setMatched', 0)
    commit('showAllCards')
    commit('setStage', 'game')
  
    dispatch('commitDelayed', {
      mutation: 'maskAllCards',
      delay: config.initialVisibleTime,
      id: 'startGame-1'
    })

    //for future
    preloadImages(['/static/cards/back.png', '/static/banners/EndGame@2x.png'])
  } else {
    commit('setStage', 'loading')
    return new Promise((resolve, reject) => {
      getters.deck.loading
        .then(() => {
          dispatch('startGame')
          resolve()
        })
        .catch(reject)
    })
  }
}


export const restartGame = ({commit, dispatch}) => {
  dispatch('startGame')
}


export const endGame = ({commit}) => {
  commit('setStage', 'end')
}


export const flipCard = ({commit, dispatch, getters}, card) => {
  commit('applyHiding')
  if((card.flipped && getters.flipped.length !== 2) || card.hidden) return

  if(getters.flipped.length === 1) {
    commit('flipCard', card)
    dispatch('matchCards')
  } else {
    commit('maskAllCards')
    commit('flipCard', card)
  }
}


export const matchCards = ({commit, dispatch, getters}) => {
  if(getters.flipped[0].id === getters.flipped[1].id) {
    dispatch('applyCardMatch')
  } else {
    const diff = config.calcScore.failure(getters.matched, getters.unmatched)
    commit('setScore', getters.score + diff)
  }
}


export const applyCardMatch = ({commit, dispatch, getters}) => {
  const diff = config.calcScore.success(getters.matched, getters.unmatched)
  commit('setScore', getters.score + diff)
  
  commit('setMatched', getters.matched + 2)

  getters.flipped.forEach(card => commit('queueCardToHide', card))

  if(getters.unmatched === 0) {
    dispatch('endGame')
  } else {
    dispatch('commitDelayed', {
      mutation: 'applyHiding',
      delay: config.matchedVisibleTime,
      id: 'applyCardMatch-1'
    })
  }
}


export const commitDelayed = ({commit}, {mutation, delay, id, payload}) => {
  if(typeof id === 'undefined') {
    console.warn(
      `Not passing id parameter to commitDelayed puts you at risk
       of unknowingly resetting needed timeout`
    )
  }


  commit('delayMutation', {
    name: `${mutation}---${id}`,
    cb: () => commit(mutation, payload), // mutations cannot commit
    delay
  })
}