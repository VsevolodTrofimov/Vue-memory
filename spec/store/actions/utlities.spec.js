import {snapAction, useMockPromise, useRealPromise, resetTimetable} from 'vuex-snapshot'

import * as actions from '@/src/store/actions.js'
import config from '@/src/utility/gameConfig'

describe('Actions', () => {
  describe('Utility actions', () => {
    describe('Load deck', () => {
      const configRealOptions = config.cards.options

      beforeEach(() => {
        config.cards.options = ['/static/card1.png', '/static/card2.png']
        useMockPromise()
        resetTimetable()
      })

      afterEach(() => {
        useRealPromise()
      })
      
      afterAll(() => {
        config.cards.options = configRealOptions
      })

      it('Matches "good loading" snapshot', done => {
        snapAction(actions.loadDeck, ['Promise', 'Promise'])
          .then(run => {
            expect(run).toMatchSnapshot()
            done()
          })
      })

      it('Matches "network error" snapshot', done => {
        snapAction(actions.loadDeck, ['Promise', {name: 'Promise', type: 'reject', payload: 403}])
          .then(run => {
            expect(run).toMatchSnapshot()
            done()
          })
      })
    })


    describe('Commit Delayed', () => {
      it('Matches snapshot', () => {
        const commit = jest.fn((mutation, payload) => {
          if(payload.cb) payload.cb()
        })
        const mutation = 'sample', payload = 'samplePayload'

        expect(snapAction(actions.commitDelayed, {
          payload: {
            delay: 150,
            id: 'test-1',
            mutation, payload
          },
          commit
        })).toMatchSnapshot()

        expect(commit).toMatchSnapshot()
      })

      it('Warns about not using id', () => {
        const realWarn = console.warn
        console.warn = jest.fn()
        
        actions.commitDelayed({commit: jest.fn()}, {})
        expect(console.warn).toBeCalled()

        console.warn = realWarn
      })
    })
  })
})