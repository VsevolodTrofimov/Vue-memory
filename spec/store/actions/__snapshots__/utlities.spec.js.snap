// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utility actions Commit Delayed Matches snapshot 1`] = `
"COMMIT: delayMutation
{\\"name\\":\\"sample---test-1\\",\\"delay\\":150}
---
COMMIT: sample
\\"samplePayload\\"
---
"
`;

exports[`Utility actions Commit Delayed Matches snapshot 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "delayMutation",
      Object {
        "cb": [Function],
        "delay": 150,
        "name": "sample---test-1",
      },
    ],
    Array [
      "sample",
      "samplePayload",
    ],
  ],
}
`;

exports[`Utility actions Load deck Matches "good loading" snapshot 1`] = `
"COMMIT: deckLoaderRegister
{}
---
RESOLUTION: Promise -> resolve
---
RESOLUTION: Promise -> resolve
---
COMMIT: deckLoaded
---
ACTION RESOLVED
---
"
`;

exports[`Utility actions Load deck Matches "network error" snapshot 1`] = `
"COMMIT: deckLoaderRegister
{}
---
RESOLUTION: Promise -> resolve
---
RESOLUTION: Promise -> reject
403
---
ACTION REJECTED
403
---
"
`;
